swift/Sources/SignalClient/Logging.m

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright 2020 Signal Messenger, LLC
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+#import "signal_ffi.h"
+#import <SignalCoreKit/OWSLogs.h>
+
+static bool isEnabled(const char *_Nonnull target, SignalLogLevel level)
+{
+    switch (level) {
+        case SignalLogLevel_Error:
+            return ShouldLogError();
+        case SignalLogLevel_Warn:
+            return ShouldLogWarning();
+        case SignalLogLevel_Info:
+            return ShouldLogInfo();
+        case SignalLogLevel_Debug:
+            return ShouldLogDebug();
+        case SignalLogLevel_Trace:
+            return ShouldLogVerbose();
+        default:
+            return ShouldLogError();
+    }
+}
+
+static void logMessage(const char *_Nonnull target,
+    SignalLogLevel level,
+    const char *_Nullable modulePath,
+    uint32_t line,
+    const char *_Nonnull message)
+{
+    // We're not using OWSLog* directly because we don't want log() to be the source of the log.
+    NSString *formattedMessage;
+    if (modulePath) {
+        formattedMessage = [NSString stringWithFormat:@"[%s:%u] %s", modulePath, line, message];
```

Comment:
Maybe this is a waste of time, but there might be a perf benefit to exiting early _before_ `[NSString stringWithFormat:...]` if we're not going to log this `level`. There's functions like `ShouldLogVerbose()` and we could add a `ShouldLogLevel(level)` function too.

Reply:
I guess it would be easy enough to call `isEnabled` before doing any other work! Some of the Rust examples even do it that way.